examples/trusted_database/module_1/rust/src/lib.rs

```diff
@@ -0,0 +1,144 @@
+pub struct TrustedDatabaseHandlerNode {
+    points_of_interest: Vec<PointOfInterest>,
+}
+
+impl TrustedDatabaseHandlerNode {
+    /// Returns a distance (in kilometers) between two locations using the Haversine formula
+    /// (ignoring height variations):
+    /// https://en.wikipedia.org/wiki/Haversine_formula
+    fn distance(first: Location, second: Location) -> f32 {
+        let first_latitude_radians = first.latitude.to_radians();
+        let second_latitude_radians = second.latitude.to_radians();
+
+        let latitude_difference_radians = (first.latitude - second.latitude).to_radians();
+        let longitude_difference_radians = (first.longitude - second.longitude).to_radians();
+
+        let central_angle = 2.0
+            * ((latitude_difference_radians / 2.0).sin().powi(2)
+                + (first_latitude_radians.cos()
+                    * second_latitude_radians.cos()
+                    * (longitude_difference_radians / 2.0).sin().powi(2)))
+            .sqrt()
+            .asin();
+
+        // Earth radius in kilometers.
+        let earth_radius = 6371.0_f32;
+
+        // Compute distance.
+        earth_radius * central_angle
+    }
+}
+
+/// A gRPC service implementation for the Private Information Retrieval example.
+impl TrustedDatabase for TrustedDatabaseHandlerNode {
+    /// Find the nearest Point Of Interest based on linear search in the database.
+    fn list_points_of_interest(
+        &mut self,
+        request: ListPointsOfInterestRequest,
+    ) -> grpc::Result<ListPointsOfInterestResponse> {
+        debug!("Received request: {:?}", request);
+        let location = request.location.ok_or_else(|| {
+            let err = "Location is not specified".to_string();
+            error!("Bad request: {:?}", err);
+            grpc::build_status(grpc::Code::InvalidArgument, &err)
+        })?;
+        let nearest_point = self.points_of_interest.iter().fold(
+            (None, f32::MAX),
+            |(closest_point, min_distance), point| {
+                let point_location = point.location.clone().expect("Non-existing location");
+                let distance = Self::distance(location.clone(), point_location);
+                if distance < min_distance {
+                    (Some(point.clone()), distance)
+                } else {
+                    (closest_point, min_distance)
+                }
+            },
+        );
+        match nearest_point.0 {
+            Some(point) => {
+                debug!("Found the nearest Point Of Interest: {:?}", point);
+                Ok(ListPointsOfInterestResponse {
+                    point_of_interest: Some(point),
+                })
+            }
+            None => {
+                let err = "Empty database".to_string();
+                error!("Couldn't find the nearest Point Of Interest: {:?}", err);
+                Err(grpc::build_status(grpc::Code::InvalidArgument, &err))
```

Comment:
nit: if the database is empty returning an InvalidArgument error to the client is confusing. Something like an internal error would be more accurate, as there is nothing that the client can do (by changing the input argument) to fix this.